src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
-            var typedRHS = GenerateConversionForAssignment(lhsAsTuple, boundRHS, diagnostics);
+            Debug.Assert(targetPlaceholder.Type.IsTupleType);
 
-            ImmutableArray<TypeSymbol> tupleTypes = typedRHS.Type.TupleElementTypes;
+            var tupleTypes = targetPlaceholder.Type.TupleElementTypes;
+            if (variables.Length != tupleTypes.Length)
+            {
+                Error(diagnostics, ErrorCode.ERR_DeconstructWrongCardinality, syntax, tupleTypes.Length, variables.Length);
+                return null;
+            }
 
-            // figure out the pairwise conversions
-            var assignments = checkedVariables.SelectAsArray((variable, index, types) => MakeAssignmentInfo(variable, types[index], node, diagnostics), tupleTypes);
+            var deconstructionStep = new DeconstructStep()
+            {
+                DeconstructMemberOpt = null,
+                TargetPlaceholder = targetPlaceholder,
+                OutputPlaceholders = tupleTypes.SelectAsArray((t, s) => new BoundDeconstructValuePlaceholder(s, t), syntax)
+            };
 
-            TypeSymbol voidType = GetSpecialType(SpecialType.System_Void, diagnostics, node);
-            return new BoundDeconstructionAssignmentOperator(node, checkedVariables, typedRHS, deconstructMemberOpt: null, assignments: assignments, type: voidType);
+            return deconstructionStep;
         }
 
-        private BoundExpression BindDeconstructWithDeconstruct(AssignmentExpressionSyntax node, DiagnosticBag diagnostics, BoundExpression boundRHS, ImmutableArray<BoundExpression> checkedVariables)
+        /// <summary>
+        /// This will generate and stack appropriate deconstruction and assignment steps for a non-tuple type.
+        /// Returns null if there was an error (if a suitable Deconstruct method was not found).
+        /// </summary>
+        static private DeconstructStep MakeNonTupleDeconstructStep(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
+                        DiagnosticBag diagnostics,
+                        ImmutableArray<BoundExpression> variables,
+                        ArrayBuilder<DeconstructStep> deconstructionSteps,
+                        ArrayBuilder<AssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
-            DiagnosticBag bag = new DiagnosticBag();
-            MethodSymbol deconstructMethod = FindDeconstruct(checkedVariables, boundRHS, node, bag);
+            var bag = DiagnosticBag.GetInstance();
+            MethodSymbol deconstructMethod = FindDeconstruct(variables.Length, targetPlaceholder, syntax, bag);
             if (!diagnostics.HasAnyErrors())
```

Comment:
I think we shouldn't be suppressing diagnostics based on whether there are other errors in the bag, unless we are controlling what might be in the bag. We can suppress additional diagnostics based on the fact that we reported something already, that was definitely related to the things we are going to be doing next.